### PR TITLE
Io improvements

### LINKED ISSUE #3174: `bitcoin_io::Cursor` is almost useless
In `std` this type can be used for seeking but here it cannot because the position cannot be modified by downstream consumers. So effectively the only use of it is to wrap owning types such as `Vec<u8>` or `Box<u8>` to get an owned type.

We should add a method to set the position. I think the `Seek` trait is kinda too much but I'm open to be convinced it's justified.

Also `position` returns `u64` which is weird given the cursor is holding a slice with max length being `usize::MAX`.

### LINKED ISSUE #3175: `bitcoin-io` is missing bridging types for `std`
We need two wrapper types to bridge with `std` - one for `std` -> `bitcoin-io`, the other in the opposite direction. We should've added them when we've removed the blanket impl but we forgot.

### PR DESCRIPTION
**Pull request by Kixunil**
This addresses significant API holes. Originally I wanted to add a commit that bumps the IO version but then I remembered we have #3162 which we most likely also want to merge and release.

Closes #3174 
Closes #3175

### CODE REVIEW

**Comment by Kixunil** [comment]
@apoelstra I did think about that but since I already had `*inner` methods etc and there's a tiny chance we one day decide to add `PhantomData` (totally hypothetical) then I might as well keep the encapsulated design. `inner` is slightly better than `0` even though not great. (But we don't have much better name in cases like this.)

I object to `Deref` because it's an anti-pattern. These types are not smart pointers.

The macro is the one implementing both traits in one go.

**Comment by apoelstra** [comment]
> I object to `Deref` because it's an anti-pattern. These types are not smart pointers.

I don't feel strongly enough to argue, but the wrapped types are conceptually identical to the unwrapped types. The wrapper exists purely to deal with Rust language details that force us to duplicate `std` traits. So `Deref` is appropriate here. But sure, it's not much extra work to write `.inner()` especially given that these types are intended to be used ephemerally.

>The macro is the one implementing both traits in one go.

I'm even more confused. We have two types, `FromStd` and `ToStd`, both of which have several distinct manually-written trait implementations. There is no macro, it's hard to imagine what a macro would be used for, and users do not need to implement any traits. Also the commit message refers to lacking "convenience methods" but I don't know what the "convenience methods" would be either.

**Comment by apoelstra** [comment]
I don't think it's worth the complexity and incompleteness of macro-generated code just to write `my_io_object.into_std()` instead of `FromStd::new(my_io_object)`. I wouldn't bother with the macros. (Though I don't mind if they exist, and if you want to use them.)

**Comment by Kixunil** [comment]
That's not what the macros are doing. They implement both `std` and our traits on the type directly, so you don't need to call `.into_std` at all. That's very nice and useful. The `ToStd` bridge is mainly helpful if you're consuming a library that didn't impl `std` traits for some reason.

**Comment by Kixunil** [comment]
I've also added the remaining impls which was requested elsewhere.

**Review by apoelstra — APPROVED** [review]
ACK 56b19d0601557509e6deb33d1b1e10208146f70b successfully ran local tests

**Comment by apoelstra** [comment]
Though for the record I still have no idea what the commit message of fc7e213f216ca91b40164a37611c0cab5633803a is supposed to mean.

**Review by tcharding — APPROVED** [review]
What was your reasoning on why not to add `Seek` trait? I believe LDK would like to use it.

I'm acking but have one nit and one question re naming.

ACK 56b19d0601557509e6deb33d1b1e10208146f70b

File: io/src/bridge.rs [hunk 1]
```diff
@@ -0,0 +1,541 @@
...
+
+    /// Returns the wrapped value.
+    #[inline]
+    pub fn into_inner(self) -> T {
+        self.0
+    }
+
+    /// Returns a reference to the wrapped value.
+    #[inline]
+    pub fn inner(&self) -> &T {
```
[the remark this answers is not in the document]
Comment (Kixunil) [reply io/src/bridge.rs:21]: I thought of it more as a getter rather than conversion.

[the remark this answers is not in the document]
Comment (tcharding) [reply io/src/bridge.rs:21]: fair enough

**Comment by Kixunil** [comment]
I wasn't under the impression that `Seek` is super interesting at least not for `bitcoin` and its dependencies. IDK what LDK people are doing with seeking.

**Comment by tcharding** [comment]
We should back port this to `io v0.1.3` so that we get it in `bitcoin v.032` - do we need to explicitly bump the `io` dependency in `bitcoin` and then do a point release of `bitcoin`?

**Comment by Kixunil** [comment]
> We should back port this to `io v0.1.3` so that we get it in `bitcoin v.032`

Oh, I thought there were no changes to `io` since then. I'll look into it.

> do we need to explicitly bump the `io` dependency in `bitcoin` and then do a point release of `bitcoin`?

No, that'd be just useless churn.

**Pull request merged**